File: backend/src/server/routes/temporal.ts
import { Elysia, t } from "elysia";
import { log } from "../../core/log";
import {
    create_fact, get_facts, create_edge, get_edges, get_related_facts,
    search_facts, get_subject_timeline, get_changes_in_window,
    get_change_frequency, compare_time_points, get_volatile_facts,
    invalidate_fact
} from "../../memory/temporal";

export const temporal = (app: Elysia) =>
    app.group("/api/temporal", (app) =>
        app
            .post("/fact", async ({ body, set }) => {
                const b = body;
                try {
                    const id = await create_fact(
                        b.subject,
                        b.predicate,
                        b.object,
                        b.valid_from,
                        b.valid_to,
                        b.confidence,
                        b.metadata
                    );
                    return { ok: true, id };
                } catch (error: any) {
                    log.error('Temporal fact creation failed', { error: error.message });
                    set.status = 500;
                    return { error: error.message };
                }
            }, {
                body: t.Object({
                    subject: t.String({ minLength: 1 }),
                    predicate: t.String({ minLength: 1 }),
                    object: t.String({ minLength: 1 }),
                    valid_from: t.Optional(t.Numeric()),
                    valid_to: t.Optional(t.Numeric()),
                    confidence: t.Optional(t.Numeric()),
                    metadata: t.Optional(t.Any())
                })
            })
            .get("/fact", async ({ query, set }) => {
                try {
                    const filters = {
                        subject: query.subject,
                        predicate: query.predicate,
                        object: query.object,
                        valid_at: query.valid_at ? Number(query.valid_at) : undefined
                    };
                    const facts = await get_facts(filters);
                    return { facts };
                } catch (error: any) {
                    log.error('Temporal fact query failed', { error: error.message });
                    set.status = 500;
                    return { error: error.message };
                }
            }, {
                query: t.Object({
                    subject: t.Optional(t.String()),
                    predicate: t.Optional(t.String()),
                    object: t.Optional(t.String()),
                    valid_at: t.Optional(t.Union([t.String(), t.Numeric()]))
                })
            })
            .delete("/fact/:id", async ({ params, query, set }) => {
                try {
                    await invalidate_fact(params.id, query.valid_to ? Number(query.valid_to) : undefined);
                    return { ok: true };
                } catch (error: any) {
                    log.error('Temporal fact invalidation failed', { error: error.message });
                    set.status = 500;
                    return { error: error.message };
                }
            }, {
                params: t.Object({
                    id: t.String()
                }),
                query: t.Object({
                    valid_to: t.Optional(t.Union([t.String(), t.Numeric()]))
                })
            })
            .post("/edge", async ({ body, set }) => {
                const b = body;
                try {
                    const id = await create_edge(
                        b.source_id,
                        b.target_id,
                        b.relation,
                        b.weight,
                        b.metadata
                    );
                    return { ok: true, id };
                } catch (error: any) {
                    log.error('Temporal edge creation failed', { error: error.message });
                    set.status = 500;
                    return { error: error.message };
                }
            }, {
                body: t.Object({
                    source_id: t.String(),
                    target_id: t.String(),
                    relation: t.String(),
                    weight: t.Optional(t.Numeric()),
                    metadata: t.Optional(t.Any())
                })
            })
            .get("/edge", async ({ query, set }) => {
                try {
                    const edges = await get_edges(query.source_id);
                    return { edges };
                } catch (error: any) {
                    log.error('Temporal edge query failed', { error: error.message });
                    set.status = 500;
                    return { error: error.message };
                }
            }, {
                query: t.Object({
                    source_id: t.String()
                })
            })
            .get("/related", async ({ query, set }) => {
                try {
                    const related = await get_related_facts(
                        query.fact_id,
                        query.relation_type,
                        query.at ? Number(query.at) : undefined
                    );
                    return { related };
                } catch (error: any) {
                    set.status = 500;
                    return { error: error.message };
                }
            }, {
                query: t.Object({
                    fact_id: t.String(),
                    relation_type: t.Optional(t.String()),
                    at: t.Optional(t.Union([t.String(), t.Numeric()]))
                })
            })
            .get("/search", async ({ query, set }) => {
                try {
                    const facts = await search_facts(
                        query.pattern,
                        query.field as any,
                        query.at ? Number(query.at) : undefined
                    );
                    return { facts };
                } catch (error: any) {
                    set.status = 500;
                    return { error: error.message };
                }
            }, {
                query: t.Object({
                    pattern: t.String(),
                    field: t.Optional(t.String()),
                    at: t.Optional(t.Union([t.String(), t.Numeric()]))
                })
            })
            .get("/timeline", async ({ query, set }) => {
                try {
                    const timeline = await get_subject_timeline(query.subject, query.predicate);
                    return { timeline };
                } catch (error: any) {
                    set.status = 500;
                    return { error: error.message };
                }
            }, {
                query: t.Object({
                    subject: t.String(),
                    predicate: t.Optional(t.String())
                })
            })
            .get("/changes", async ({ query, set }) => {
                try {
                    const timeline = await get_changes_in_window(
                        Number(query.start),
                        Number(query.end),
                        query.subject
                    );
                    return { timeline };
                } catch (error: any) {
                    set.status = 500;
                    return { error: error.message };
                }
            }, {
                query: t.Object({
                    start: t.Union([t.String(), t.Numeric()]),
                    end: t.Union([t.String(), t.Numeric()]),
                    subject: t.Optional(t.String())
                })
            })
            .get("/frequency", async ({ query, set }) => {
                try {
                    const result = await get_change_frequency(
                        query.subject,
                        query.predicate,
                        query.window_days ? Number(query.window_days) : undefined
                    );
                    return { result };
                } catch (error: any) {
                    set.status = 500;
                    return { error: error.message };
                }
            }, {
                query: t.Object({
                    subject: t.String(),
                    predicate: t.String(),
                    window_days: t.Optional(t.Union([t.String(), t.Numeric()]))
                })
            })
            .get("/compare", async ({ query, set }) => {
                try {
                    const result = await compare_time_points(
                        query.subject,
                        Number(query.time1),
                        Number(query.time2)
                    );
                    return { result };
                } catch (error: any) {
                    set.status = 500;
                    return { error: error.message };
                }
            }, {
                query: t.Object({
                    subject: t.String(),
                    time1: t.Union([t.String(), t.Numeric()]),
                    time2: t.Union([t.String(), t.Numeric()])
                })
            })
            .get("/volatile", async ({ query, set }) => {
                try {
                    const result = await get_volatile_facts(
                        query.subject,
                        query.limit ? Number(query.limit) : undefined
                    );
                    return { result };
                } catch (error: any) {
                    set.status = 500;
                    return { error: error.message };
                }
            }, {
                query: t.Object({
                    subject: t.Optional(t.String()),
                    limit: t.Optional(t.Union([t.String(), t.Numeric()]))
                })
            })
    );

File: packages/openmemory-py/src/openmemory/server/routers/temporal.py
from fastapi import APIRouter, Depends, HTTPException
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field, ConfigDict
from ...main import Memory
from ...core.types import TemporalFact
from ..dependencies import get_current_user_id, resolve_user

router = APIRouter()
mem_client = Memory()

class FactRequest(BaseModel):
    subject: str
    predicate: str
    fact_object: str = Field(..., alias="object")
    valid_from: Optional[str] = Field(None, alias="validFrom")
    confidence: float = 1.0
    user_id: Optional[str] = Field(None, alias="userId")
    metadata: Dict[str, Any] = {}
    
    model_config = ConfigDict(populate_by_name=True)

class EdgeRequest(BaseModel):
    source_id: str = Field(..., alias="sourceId")
    target_id: str = Field(..., alias="targetId")
    relation_type: str = Field(..., alias="relationType")
    weight: float = 1.0
    user_id: Optional[str] = Field(None, alias="userId")

    model_config = ConfigDict(populate_by_name=True)

@router.post("/fact", response_model_by_alias=True)
async def create_fact(req: FactRequest, auth_user: str = Depends(get_current_user_id)):
    t = mem_client.temporal  # type: ignore[attr-defined]
    from ...temporal_graph import insert_fact
    uid = resolve_user(auth_user, req.user_id)
    valid_from_val = None
    if req.valid_from:
        if req.valid_from.isdigit():
            valid_from_val = int(req.valid_from)
        else:
            try:
                from dateutil import parser
                dt = parser.parse(req.valid_from)
                valid_from_val = int(dt.timestamp() * 1000)
            except Exception as e:
                import logging
                logging.getLogger("temporal").warning(f"Failed to parse date {req.valid_from}: {e}")

    fid = await insert_fact(
        req.subject, req.predicate, req.fact_object, 
        valid_from=valid_from_val,  # type: ignore[call-arg]
        confidence=req.confidence,
        metadata=req.metadata,
        user_id=uid  # type: ignore[call-arg]
    )
    return {"id": fid}

@router.get("/fact", response_model_by_alias=True)
async def get_facts_filtered(
    subject: Optional[str] = None,
    predicate: Optional[str] = None,
    object: Optional[str] = None,
    at: Optional[int] = None,
    minConfidence: Optional[float] = None,
    limit: int = 100,
    user_id: Optional[str] = None,
    auth_user: str = Depends(get_current_user_id)
):
    from ...temporal_graph import query_facts_at_time
    uid = resolve_user(auth_user, user_id)
    
    facts = await query_facts_at_time(
        subject=subject,
        predicate=predicate,
        fact_object=object,
        at=at,
        min_confidence=minConfidence or 0.0,  # type: ignore[call-arg]
        user_id=uid  # type: ignore[call-arg]
    )
    return {"facts": facts}

@router.get("/search", response_model_by_alias=True)
async def search_facts_pattern(
    pattern: Optional[str] = None, 
    q: Optional[str] = None,  # Support both
    type: str = "subject", 
    limit: int = 10, 
    user_id: Optional[str] = None, 
    auth_user: str = Depends(get_current_user_id)
):
    from ...temporal_graph import search_facts
    uid = resolve_user(auth_user, user_id)
    query_str = pattern or q
    if not query_str:
        return {"facts": []}

    # search_facts in query.py currently only supports pattern search (likely subject based)
    res = await search_facts(query_str, limit=limit, user_id=uid)  # type: ignore[call-arg]
    return {"facts": res}

@router.post("/edge", response_model_by_alias=True)
async def create_edge(req: EdgeRequest, auth_user: str = Depends(get_current_user_id)):
    from ...temporal_graph.store import insert_edge
    uid = resolve_user(auth_user, req.user_id)
    eid = await insert_edge(
        req.source_id, req.target_id, req.relation_type,
        weight=req.weight, user_id=uid  # type: ignore[call-arg]
    )
    return {"id": eid, "ok": True}

@router.get("/edge", response_model_by_alias=True)
async def get_edges_route(sourceId: Optional[str] = None, targetId: Optional[str] = None, relationType: Optional[str] = None, user_id: Optional[str] = None, auth_user: str = Depends(get_current_user_id)):
    from ...temporal_graph import query_edges
    uid = resolve_user(auth_user, user_id)
    edges = await query_edges(source_id=sourceId, target_id=targetId, relation_type=relationType, user_id=uid)  # type: ignore[call-arg]
    return {"edges": edges}

@router.get("/subject/{subject}", response_model_by_alias=True)
async def get_subject_facts(subject: str, user_id: Optional[str] = None, auth_user: str = Depends(get_current_user_id)):
    from ...temporal_graph.query import get_facts_by_subject
    uid = resolve_user(auth_user, user_id)
    facts = await get_facts_by_subject(subject, user_id=uid)  # type: ignore[call-arg]
    return {"facts": facts}

@router.get("/timeline", response_model_by_alias=True)
async def get_timeline_route(subject: str, user_id: Optional[str] = None, auth_user: str = Depends(get_current_user_id)):
    from ...temporal_graph import get_subject_timeline
    uid = resolve_user(auth_user, user_id)
    timeline = await get_subject_timeline(subject, user_id=uid)  # type: ignore[call-arg]
    return {"timeline": timeline}

@router.get("/history/predicate", response_model_by_alias=True)
async def get_predicate_history_route(predicate: str, user_id: Optional[str] = None, auth_user: str = Depends(get_current_user_id)):
    from ...temporal_graph import get_predicate_timeline
    uid = resolve_user(auth_user, user_id)
    timeline = await get_predicate_timeline(predicate, user_id=uid)  # type: ignore[call-arg]
    return {"timeline": timeline}
